File: app/src/main/java/com/ssy/ferry/core/MethodMonitor.java
package com.ssy.ferry.core;

import android.app.Activity;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.os.SystemClock;

import com.ssy.ferry.util.FerryLog;

import java.util.HashSet;
import java.util.Set;

public class MethodMonitor extends MonitorLifecycle {

    private static final String TAG = "Ferry.MethodMonitor";
    private static MethodMonitor sInstance = new MethodMonitor();
    private static final int STATUS_DEFAULT = Integer.MAX_VALUE;
    private static final int STATUS_STARTED = 2;
    private static final int STATUS_READY = 1;
    private static final int STATUS_STOPPED = -1;
    private static final int STATUS_EXPIRED_START = -2;
    private static final int STATUS_OUT_RELEASE = -3;

    private static volatile int status = STATUS_DEFAULT;
    private static Object statusLock = new Object();

    private static long[] sBuffer = new long[Constants.BUFFER_SIZE];
    private static int sIndex = 0;
    private static int sLastIndex = -1;
    private static boolean assertIn = false;
    private volatile static long sCurrentDiffTime = SystemClock.uptimeMillis();
    private volatile static long sDiffTime = sCurrentDiffTime;
    private static Thread sMainThread = Looper.getMainLooper().getThread();
    private static HandlerThread sTimerUpdateThread = FerryHandlerThread.getNewHandlerThread("matrix_time_update_thread");
    private static Handler sHandler = new Handler(sTimerUpdateThread.getLooper());
    private static final int METHOD_ID_MAX = 0xFFFFF;
    public static final int METHOD_ID_DISPATCH = METHOD_ID_MAX - 1;
    private static Set<String> sFocusActivitySet = new HashSet<>();
    private static String sFocusedActivity = "default";
    private static HashSet<IMethodMonitorListener> listeners = new HashSet<>();
    private static Object updateTimeLock = new Object();
    private static boolean isPauseUpdateTime = false;
    private static Runnable checkStartExpiredRunnable = null;

    static {
        sHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                realRelease();
            }
        }, Constants.DEFAULT_RELEASE_BUFFER_DELAY);
    }

    /**
     * update time runnable
     */
    private static Runnable sUpdateDiffTimeRunnable = new Runnable() {
        @Override
        public void run() {
            try {
                while (true) {
                    while (!isPauseUpdateTime && status > STATUS_STOPPED) {
                        sCurrentDiffTime = SystemClock.uptimeMillis() - sDiffTime;
                        SystemClock.sleep(Constants.TIME_UPDATE_CYCLE_MS);
                    }
                    synchronized (updateTimeLock) {
                        updateTimeLock.wait();
                    }
                }
            } catch (InterruptedException e) {
                FerryLog.e(TAG, "" + e.toString());
            }
        }
    };

    public static MethodMonitor getInstance() {
        return sInstance;
    }


    public boolean isAlive() {
        return status >= STATUS_STARTED;
    }


    public static boolean isRealTrace() {
        return status >= STATUS_READY;
    }

    private static void realRelease() {
        synchronized (statusLock) {
            if (status == STATUS_DEFAULT) {
                FerryLog.i(TAG, "[realRelease] timestamp:%s", System.currentTimeMillis());
                sHandler.removeCallbacksAndMessages(null);
                sTimerUpdateThread.quit();
                sBuffer = null;
                status = STATUS_OUT_RELEASE;
            }
        }
    }

    private static void realExecute() {
        FerryLog.i(TAG, "[realExecute] timestamp:%s", System.currentTimeMillis());

        sCurrentDiffTime = SystemClock.uptimeMillis() - sDiffTime;

        sHandler.removeCallbacksAndMessages(null);
        sHandler.postDelayed(sUpdateDiffTimeRunnable, Constants.TIME_UPDATE_CYCLE_MS);
        sHandler.postDelayed(checkStartExpiredRunnable = new Runnable() {
            @Override
            public void run() {
                synchronized (statusLock) {
                    if (status == STATUS_DEFAULT || status == STATUS_READY) {
                        status = STATUS_EXPIRED_START;
                    }
                }
            }
        }, Constants.DEFAULT_RELEASE_BUFFER_DELAY);

      ActivityThreadHacker.hackSysHandlerCallback();
        LooperMonitor.register(new LooperMonitor.LooperDispatchListener() {
            @Override
            public boolean isValid() {
                return status >= STATUS_READY;
            }

            @Override
            public void dispatchStart() {
                super.dispatchStart();
                MethodMonitor.dispatchBegin();
            }

            @Override
            public void dispatchEnd() {
                super.dispatchEnd();
                MethodMonitor.dispatchEnd();
            }
        });
    }

    private static void dispatchBegin() {
        sCurrentDiffTime = SystemClock.uptimeMillis() - sDiffTime;
        isPauseUpdateTime = false;

        synchronized (updateTimeLock) {
            updateTimeLock.notifyAll();
        }
    }

    private static void dispatchEnd() {
        isPauseUpdateTime = true;
    }


    /**
     * hook method when it's called in.
     *
     * @param methodId
     */
    public static void i(int methodId) {

        if (status <= STATUS_STOPPED) {
            return;
        }
        if (methodId >= METHOD_ID_MAX) {
            return;
        }

        if (status == STATUS_DEFAULT) {
            synchronized (statusLock) {
                if (status == STATUS_DEFAULT) {
                    realExecute();
                    status = STATUS_READY;
                }
            }
        }

        if (Thread.currentThread() == sMainThread) {
            if (assertIn) {
                android.util.Log.e(TAG, "ERROR!!! MethodMonitor.i Recursive calls!!!");
                return;
            }
            assertIn = true;
            if (sIndex < Constants.BUFFER_SIZE) {
                mergeData(methodId, sIndex, true);
            } else {
                sIndex = -1;
            }
            ++sIndex;
            assertIn = false;
        }
    }

    /**
     * hook method when it's called out.
     *
     * @param methodId
     */
    public static void o(int methodId) {

        if (status <= STATUS_STOPPED) {
            return;
        }
        if (methodId >= METHOD_ID_MAX) {
            return;
        }
        if (Thread.currentThread() == sMainThread) {
            if (sIndex < Constants.BUFFER_SIZE) {
                mergeData(methodId, sIndex, false);
            } else {
                sIndex = -1;
            }
            ++sIndex;
        }
    }

    /**
     * when the special method calls,it's will be called.
     *
     * @param activity now at which activity
     * @param isFocus  this window if has focus
     */
    public static void at(Activity activity, boolean isFocus) {
        String activityName = activity.getClass().getName();
        if (isFocus) {
            sFocusedActivity = activityName;
            if (sFocusActivitySet.add(activityName)) {
                synchronized (listeners) {
                    for (IMethodMonitorListener listener : listeners) {
                        listener.onActivityFocused(activityName);
                    }
                }
                FerryLog.i(TAG, "[at] Activity[%s] has %s focus!", activityName, isFocus ? "attach" : "detach");
            }
        } else {
            if (sFocusActivitySet.remove(activityName)) {
                FerryLog.i(TAG, "[at] Activity[%s] has %s focus!", activityName, isFocus ? "attach" : "detach");
            }
        }


    }


    public static String getFocusedActivity() {
        return sFocusedActivity;
    }

    /**
     * merge trace info as a long data
     *
     * @param methodId
     * @param index
     * @param isIn
     */
    private static void mergeData(int methodId, int index, boolean isIn) {
        long trueId = 0L;
        if (isIn) {
            trueId |= 1L << 63;
        }
        trueId |= (long) methodId << 43;
        trueId |= sCurrentDiffTime & 0x7FFFFFFFFFFL;
        sBuffer[index] = trueId;
        checkPileup(index);
        sLastIndex = index;
    }

    public void addListener(IMethodMonitorListener listener) {
        synchronized (listeners) {
            listeners.add(listener);
        }
    }
    public void clearListener( ) {
        synchronized (listeners) {
            listeners.clear();
        }
    }

    public void removeListener(IMethodMonitorListener listener) {
        synchronized (listeners) {
            listeners.remove(listener);
        }
    }

    private static IndexRecord sIndexRecordHead = null;

    public IndexRecord maskIndex(String source) {
        if (sIndexRecordHead == null) {
            sIndexRecordHead = new IndexRecord(sIndex - 1);
            sIndexRecordHead.source = source;
            return sIndexRecordHead;
        } else {
            IndexRecord indexRecord = new IndexRecord(sIndex - 1);
            indexRecord.source = source;
            IndexRecord record = sIndexRecordHead;
            IndexRecord last = null;
            while (record != null) {
                if (indexRecord.index <= record.index) {
                    if (null == last) {
                        IndexRecord tmp = sIndexRecordHead;
                        sIndexRecordHead = indexRecord;
                        indexRecord.next = tmp;
                    } else {
                        IndexRecord tmp = last.next;
                        last.next = indexRecord;
                        indexRecord.next = tmp;
                    }
                    return indexRecord;
                }
                last = record;
                record = record.next;
            }

            last.next = indexRecord;

            return indexRecord;
        }
    }

    private static void checkPileup(int index) {
        IndexRecord indexRecord = sIndexRecordHead;
        while (indexRecord != null) {
            if (indexRecord.index == index || (indexRecord.index == -1 && sLastIndex == Constants.BUFFER_SIZE - 1)) {
                indexRecord.isValid = false;
                FerryLog.w(TAG, "[checkPileup] %s", indexRecord.toString());
                sIndexRecordHead = indexRecord = indexRecord.next;
            } else {
                break;
            }
        }
    }

    @Override
    public void start() {
        synchronized (statusLock) {
            if (status < STATUS_STARTED && status >= STATUS_EXPIRED_START) {
                sHandler.removeCallbacks(checkStartExpiredRunnable);
                if (sBuffer == null) {
                    throw new RuntimeException(TAG + " sBuffer == null");
                }
                //  FerryLog.i(TAG, "[onStart] preStatus:%s", status, Utils.getStack());
                this.status = STATUS_STARTED;
            } else {
                FerryLog.w(TAG, "[onStart] current status:%s", status);
            }
        }
    }

    @Override
    public void stop() {
        synchronized (statusLock) {
            if (status == STATUS_STARTED) {
                //  FerryLog.i(TAG, "[onStop] %s", Utils.getStack());
                this.status = STATUS_STOPPED;
            } else {
                FerryLog.w(TAG, "[onStop] current status:%s", status);
            }
        }
    }


    public static final class IndexRecord {
        public IndexRecord(int index) {
            this.index = index;
        }

        public IndexRecord() {
            this.isValid = false;
        }

        public int index;
        private IndexRecord next;
        public boolean isValid = true;
        public String source;

        public void release() {
            isValid = false;
            IndexRecord record = sIndexRecordHead;
            IndexRecord last = null;
            while (null != record) {
                if (record == this) {
                    if (null != last) {
                        last.next = record.next;
                    } else {
                        sIndexRecordHead = record.next;
                    }
                    record.next = null;
                    break;
                }
                last = record;
                record = record.next;
            }
        }

        @Override
        public String toString() {
            return "index:" + index + ",\tisValid:" + isValid + " source:" + source;
        }
    }

    public long[] copyData(IndexRecord startRecord) {
        return copyData(startRecord, new IndexRecord(sIndex - 1));
    }

    public long[] copyData(IndexRecord startRecord, IndexRecord endRecord) {
        long current = System.currentTimeMillis();
        long[] data = new long[0];
        try {
            if (startRecord.isValid && endRecord.isValid) {
                int length;
                int start = Math.max(0, startRecord.index);
                int end = Math.max(0, endRecord.index);

                if (end > start) {
                    length = end - start + 1;
                    data = new long[length];
                    System.arraycopy(sBuffer, start, data, 0, length);
                } else if (end < start) {
                    length = 1 + end + (sBuffer.length - start);
                    data = new long[length];
                    System.arraycopy(sBuffer, start, data, 0, sBuffer.length - start);
                    System.arraycopy(sBuffer, 0, data, sBuffer.length - start, end + 1);
                }
                return data;
            }
            return data;
        } catch (OutOfMemoryError e) {
            FerryLog.e(TAG, e.toString());
            return data;
        } finally {
            FerryLog.i(TAG, "[copyData] [%s:%s] cost:%sms", Math.max(0, startRecord.index), endRecord.index, System.currentTimeMillis() - current);
        }
    }

    public static long getDiffTime() {
        return sDiffTime;
    }

    public void printIndexRecord() {
        StringBuilder ss = new StringBuilder(" \n");
        IndexRecord record = sIndexRecordHead;
        while (null != record) {
            ss.append(record).append("\n");
            record = record.next;
        }
        FerryLog.i(TAG, "[printIndexRecord] %s", ss.toString());
    }

    public interface IMethodMonitorListener {
        void onActivityFocused(String activityName);
    }
}

File: app/src/main/java/com/ssy/ferry/core/MethodMonitor.kt
package com.ssy.ferry.core

import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.util.Log
import com.ssy.ferry.listener.LooperDispatchListener

/**
 * 2019-10-16
 * @author Mr.S
 */
class MethodMonitor : MonitorLifecycle {

    override fun start() {
        synchronized(statusLock) {
            if (status < STATUS_STARTED) {
                status = STATUS_STARTED
                alive = true
            }

        }
    }

    override fun stop() {
        synchronized(statusLock) {
            if (status > STATUS_STOPED) {
                status = STATUS_STOPED
                alive = false
            }

        }

    }

    fun copyData(startRecord: MethodMonitor.Companion.IndexRecord): LongArray {


        return copyData(startRecord, IndexRecord(mIndex - 1))
    }

    fun copyData(
        startRecord: MethodMonitor.Companion.IndexRecord,
        endRecord: MethodMonitor.Companion.IndexRecord
    ): LongArray {

        var data = LongArray(0)
        if (endRecord.index == -1) {
            return data;
        }
        try {
            if (startRecord.isValid and endRecord.isValid) {
                var length: Int
                val start = startRecord.index
                val end = endRecord.index
                if (end > start) {
                    length = end - start + 1
                    data = LongArray(length)
                    System.arraycopy(sBuffer, start, data, 0, length)
                } else { //方法满了之后 又开始从新技术  所以 end 会小于 start  所以要分两部分来复制
                    length = 1 + end + (sBuffer.size - start)
                    data = LongArray(length)
                    System.arraycopy(sBuffer, start, data, 0, sBuffer.size - start)
                    System.arraycopy(sBuffer, 0, data, sBuffer.size - start, end + 1)
                }
                return data
            }
        } catch (e: OutOfMemoryError) {

        }
        return data
    }

    companion object {
        val METHOD_ID_MAX = 0xFFFFF
      @JvmField  open var METHOD_ID_DISPATCH = METHOD_ID_MAX - 1
        private val sMainThread = Looper.getMainLooper().thread
        private var sBuffer: LongArray = LongArray(Constants.BUFFER_SIZE)
        private val statusLock = Object()
        private var mIndex = 0
        val looperDispatchListener = object : LooperDispatchListener() {
            override fun isAlive(): Boolean {
                return alive
            }

            override fun dispatchStart() {
                super.dispatchStart()

                MethodMonitor.dispatchStart()
            }

            override fun dispatchEnd() {
                super.dispatchEnd()
                MethodMonitor.dispatchEnd()
            }
        }
        private val mInstance = MethodMonitor()
        private val TAG = MethodMonitor::class.java.name
        private val STATUS_DEFAULT = Integer.MAX_VALUE
        private val STATUS_STARTED = 2
        private val STATUS_STOPED = -1
        var alive = false
        var status = STATUS_DEFAULT
        private val sTimerUpdateThread =
            FerryHandlerThread.getNewHandlerThread("matrix_time_update_thread")
        private val sHandler = Handler(sTimerUpdateThread.looper)
        private val updateTimeLock = Object()
        var isUpdateTime = false

        var curDiffTime: Long = SystemClock.uptimeMillis()
      @JvmField  var diffTime: Long = curDiffTime


        open fun i(methodId: Int) {
            if (status <= STATUS_STOPED) {
                return
            }
            if (methodId >= METHOD_ID_MAX) {
                return
            }
            curDiffTime = SystemClock.uptimeMillis() - diffTime
            if (status == STATUS_DEFAULT) {
                synchronized(statusLock) {
                    realStart()
                }
            }
            if (Thread.currentThread() === sMainThread) {
                if (mIndex < Constants.BUFFER_SIZE) {
                    flatData(methodId, mIndex, true)
                } else {
                    mIndex = -1
                }
                ++mIndex
            }
        }


        open fun o(methodId: Int) {
            if (status <= STATUS_STOPED) {
                return
            }
            if (methodId >= METHOD_ID_MAX) {
                return
            }

            if (Thread.currentThread() === sMainThread) {
                if (mIndex < Constants.BUFFER_SIZE) {
                    flatData(methodId, mIndex, false)
                } else {
                    mIndex = -1
                }
                ++mIndex
            }
        }

        /**
         * 把方法信息变成 long 类型存储 方便上传信息
         *
         */
        private fun flatData(methodId: Int, mIndex: Int, isIn: Boolean) {
            if (methodId == METHOD_ID_DISPATCH) {
                curDiffTime = SystemClock.uptimeMillis() - diffTime
            }
            var trueId = 0L
            if (isIn) {
                trueId = trueId or (1L shl 63)
            }
            trueId = trueId or (methodId.toLong() shl 43)
            trueId = trueId or (curDiffTime and 0x7FFFFFFFFFFL)
            sBuffer?.set(mIndex, trueId)


        }

        private fun realStart() {
            sHandler.removeCallbacksAndMessages(null)
            sHandler.postDelayed(updateTimeRunnable, Constants.TIME_UPDATE_CYCLE_MS)
            LooperMonitor.register(looperDispatchListener)
        }


        fun dispatchStart() {
            isUpdateTime = true
            diffTime = SystemClock.uptimeMillis() - diffTime
            synchronized(updateTimeLock) {
                updateTimeLock.notifyAll()
            }
        }

        fun dispatchEnd() {
            isUpdateTime = false
        }


        open fun getInstance(): MethodMonitor {
            return mInstance
        }

        private val updateTimeRunnable = Runnable {
            try {
                while (true) {
                    while (isUpdateTime and (status > STATUS_STOPED)) {
                        curDiffTime = SystemClock.uptimeMillis() - diffTime
                        Log.e(TAG, "curDiffTime:  $curDiffTime")
                        try {
                            Thread.sleep(Constants.TIME_UPDATE_CYCLE_MS)
                        } catch (e: InterruptedException) {

                        }

                    }
                    synchronized(updateTimeLock) {
                        updateTimeLock.wait()
                    }

                }
            } catch (e: InterruptedException) {
            }
        }
        private var mIndexRecordHead: IndexRecord? = null

        /**
         *这个的含义 就是一个链表  有一个HEAD 代表起始 END 代表最后
         */
        class IndexRecord {
            var index: Int = 0
            var isValid = true
            var source: String? = null
            var next: IndexRecord? = null

            constructor() {
                this.isValid = false
            }

            constructor(index: Int) {
                this.index = index
            }

            /**
             * 该节点 到 head 节点都删除 并赋值新的 head 节点
             */
            fun release() {
                isValid = false
                var record: IndexRecord? = mIndexRecordHead
                var last: IndexRecord? = null
                while (null != record) {
                    if (record == this) {
                        if (null != last) {
                            last.next = record.next
                        } else {
                            mIndexRecordHead = record.next
                        }
                        record.next = null
                        break
                    }
                    last = record
                    record = record.next
                }
            }

            override fun toString(): String {
                return "IndexRecord(index=$index, isValid=$isValid, source=$source)"
            }


        }

    }

    /**
     * 创建 IndexRecord
     */
    fun maskIndex(source: String): IndexRecord {

        if (mIndexRecordHead == null) {
            mIndexRecordHead = IndexRecord(mIndex - 1)
            mIndexRecordHead?.source = source
            return mIndexRecordHead as IndexRecord
        } else {
            val indexRecord = IndexRecord(mIndex - 1)
            indexRecord.source = source
            var record: IndexRecord? = mIndexRecordHead
            var last: IndexRecord? = null
            while (record != null) {
                if (indexRecord.index <= record.index) {
                    if (null == last) {
                        val tmp = mIndexRecordHead
                        mIndexRecordHead = indexRecord
                        indexRecord.next = tmp
                    } else {
                        val tmp = last.next
                        last.next = indexRecord
                        indexRecord.next = tmp
                    }
                    return indexRecord
                }
                last = record
                record = record.next
            }

            last?.next = indexRecord

            return indexRecord
        }
    }

}